hs/src/Reach/Eval/Core.hs

```diff
@@ -4291,7 +4326,28 @@ compilePayAmt tt v = do
                         return $ ((seenNet, sks'), (DLPayAmt nts $ tks <> [(amt_a, token_a)]))
                       _ -> locAt v_at $ expect_ $ Err_Token_DynamicRef
                   _ -> locAt v_at $ expect_t v $ Err_Transfer_Type tt
-          snd <$> (foldlM go ((False, []), DLPayAmt (DLA_Literal $ DLL_Int at 0) []) $ zip ts aes)
+          snd <$> (foldlM go ((False, []), mtPay) $ zip ts aes)
+        -- XXX refactor with above?
+        DLAE_Arg (DLA_Var dv) -> do
+          let go sa (ty, idx) =
+                case ty of
+                  T_UInt -> do
+                    let ((seenNet, sks), DLPayAmt _ tks) = sa
+                    when seenNet $ do
+                      locAt v_at $ expect_ $ Err_Transfer_DoubleNetworkToken tt
+                    a <- DLA_Var <$> (ctxt_lift_expr (DLVar at Nothing T_UInt) $ DLE_TupleRef at (DLA_Var dv) idx)
+                    return $ ((True, sks), DLPayAmt a tks)
+                  tupTy@(T_Tuple [T_UInt, T_Token]) -> do
+                    let ((seenNet, sks), DLPayAmt nts tks) = sa
+                    tup <- DLA_Var <$> (ctxt_lift_expr (DLVar at Nothing tupTy) $ DLE_TupleRef at (DLA_Var dv) idx)
+                    amt_a <- DLA_Var <$> (ctxt_lift_expr (DLVar at Nothing T_UInt) $ DLE_TupleRef at tup 0)
+                    token_a <- DLA_Var <$> (ctxt_lift_expr (DLVar at Nothing T_UInt) $ DLE_TupleRef at tup 1)
+                    when (token_a `elem` sks) $ do
```

Comment:
@jeapostrophe Would verifying that all the tokens in the PayAmt are unique be left up to the programmer? For example, if there are two dynamic token references, would something like the following be required:

```Javascript
const tokA = // dynamic computation
const tokB = // dynamic computation
A.pay([ [1, tokA], [1, tokB] ]).payRequire(() => {
  require(tokA != tokB);
);
```

Reply:
First, for now, you'd generate an `assert` and then the user is responsible for adding the appropriate `assume`/`require`. Meaning `Err_Transfer_DoubleToken` goes away and is replaced by a verification error

Second, ideally we'd be able to remove this constraint all together and support examples where the same token is given multiple times.